src/Avalonia.Controls/Decorator.cs

```diff
@@ -1,12 +1,23 @@
+using System;
+using Avalonia.Collections;
+using Avalonia.Controls;
+using Avalonia.Controls.Shapes;
+using Avalonia.Controls.Utils;
 using Avalonia.Layout;
+using Avalonia.Media;
 using Avalonia.Metadata;
+using Avalonia.Platform;
+using Avalonia.Rendering.Composition;
+using Avalonia.Styling;
+using Avalonia.Utilities;
+using Avalonia.VisualTree;
 
 namespace Avalonia.Controls
 {
     /// <summary>
     /// Base class for controls which decorate a single child control.
     /// </summary>
-    public class Decorator : Control
+    public class Decorator : Control, IContainer
```

Comment:
The choice to mark controls as possible containers seems to be rather arbitrary.

What was the logic behind choosing Decorator, ContentControl and ContentPresenter? Why do we have _both_ ContentControl and ContentPresenter as containers?

Reply:
Those are chosen because they are expected to have 1 child, and we can predict how it measures, thus makes more sense as containers. As for ContentControl and Presenter, that ContentPresenter should be the one container, so it is a mistake to make ControlControl one. I'll update it. 
We could make all Controls be a potential container, but that won't make sense for controls that do not expect content, like Calendar and NumberUpDown

Reply:
On second thought, even if a Control implements IContainer, it doesn't implicitly become a target for queries, as its ContainerType would be Normal. So ContentControl should be an IContainer. ContentPresenter is made to be an IContainer so allow users to make customer controls that can act as a container without implementing IContainer.

Reply:
I don't think the implementation is generic enough as-is. I see no reason a container query can't target ANY control. We can already opt-out by default with type Normal.

Keep in mind that the design of XAML is nesting of controls. That means this should be opt-out rather than opt-in. IMO the controls where it just never makes sense should have a mechanism to disable the container query provided by Control base by overriding the type.